<a href="https://colab.research.google.com/github/ulisesjcaroflores/Week2-GSB530F25/blob/main/Titanic_Wrangling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Titanic Wrangling

In this practice activity you'll continue to work with the titanic dataset in ways that flex what you've learned about both data wrangling and data visualization.

In [92]:
# Import the titanic dataset here
import numpy as np
import pandas as pd

data_dir = "https://dlsun.github.io/pods/data/"
df_titanic = pd.read_csv(data_dir + "titanic.csv")
df_titanic.head(20)

name  gender        age class  \
0                        Abbing, Mr. Anthony    male  42.000000   3rd   
1                  Abbott, Mr. Eugene Joseph    male  13.000000   3rd   
2                Abbott, Mr. Rossmore Edward    male  16.000000   3rd   
3             Abbott, Mrs. Rhoda Mary 'Rosa'  female  39.000000   3rd   
4                Abelseth, Miss. Karen Marie  female  16.000000   3rd   
5              Abelseth, Mr. Olaus Jørgensen    male  25.000000   3rd   
6                        Abelson, Mr. Samuel    male  30.000000   2nd   
7                       Abelson, Mrs. Hannah  female  28.000000   2nd   
8               Abī-Al-Munà, Mr. Nāsīf Qāsim    male  27.000000   3rd   
9   Abrahamsson, Mr. Abraham August Johannes    male  20.000000   3rd   
10                        Ådahl, Mr. Mauritz    male  30.000000   3rd   
11                           Adams, Mr. John    male  27.000000   3rd   
12            Ahlin, Mrs. Johanna Persdotter  female  40.000000   3rd   
13                 Aks, Master. Frank Philip    male   0.833333   3rd   
14                            Aks, Mrs. Leah  female  18.000000   3rd   
15              Aldworth, Mr. Augustus Henry    male  35.000000   2nd   
16             Alexander, Mr. William Albert    male  25.000000   3rd   
17               Alhomäki, Mr. Ilmari Rudolf    male  19.000000   3rd   
18                            Ali, Mr. Ahmed    male  24.000000   3rd   
19                          Ali, Mr. William    male  25.000000   3rd   

   embarked        country   ticketno     fare  survived  
0         S  United States     5547.0   7.1100         0  
1         S  United States     2673.0  20.0500         0  
2         S  United States     2673.0  20.0500         0  
3         S        England     2673.0  20.0500         1  
4         S         Norway   348125.0   7.1300         1  
5         S  United States   348122.0   7.1300         1  
6         C         France     3381.0  24.0000         0  
7         C         France     3381.0  24.0000         1  
8         C        Lebanon     2699.0  18.1509         1  
9         S        Finland  3101284.0   7.1806         1  
10        S         Sweden     7076.0   7.0500         0  
11        S        England   341826.0   8.0100         0  
12        S         Sweden     7546.0   9.0906         0  
13        S        England   392091.0   9.0700         1  
14        S        England   392091.0   9.0700         1  
15        S        England   248744.0  13.0000         0  
16        S        England     3474.0   7.1709         0  
17        S        Finland  3101287.0   7.1806         0  
18        S      Argentina  3101311.0   7.0100         0  
19        S      Argentina  3101312.0   7.0100         0

## 1. Filter the data to include passengers only. Calculate the joint distribution (cross-tab) between a passenger's class and where they embarked.

In [93]:
df_titanic["type"] = df_titanic["class"].map({
    "1st": "passenger",
    "2nd": "passenger",
    "3rd": "passenger",
    "victualling crew": "crew",
    "engineering crew": "crew",
    "deck crew": "crew"
})

df_titanic

name  gender   age             class embarked  \
0                Abbing, Mr. Anthony    male  42.0               3rd        S   
1          Abbott, Mr. Eugene Joseph    male  13.0               3rd        S   
2        Abbott, Mr. Rossmore Edward    male  16.0               3rd        S   
3     Abbott, Mrs. Rhoda Mary 'Rosa'  female  39.0               3rd        S   
4        Abelseth, Miss. Karen Marie  female  16.0               3rd        S   
...                              ...     ...   ...               ...      ...   
2202                Wynn, Mr. Walter    male  41.0         deck crew        B   
2203             Yearsley, Mr. Harry    male  40.0  victualling crew        S   
2204        Young, Mr. Francis James    male  32.0  engineering crew        S   
2205             Zanetti, Sig. Minio    male  20.0  restaurant staff        S   
2206              Zarracchi, Sig. L.    male  26.0  restaurant staff        S   

            country  ticketno   fare  survived       type  
0     United States    5547.0   7.11         0  passenger  
1     United States    2673.0  20.05         0  passenger  
2     United States    2673.0  20.05         0  passenger  
3           England    2673.0  20.05         1  passenger  
4            Norway  348125.0   7.13         1  passenger  
...             ...       ...    ...       ...        ...  
2202        England       NaN    NaN         1       crew  
2203        England       NaN    NaN         1       crew  
2204        England       NaN    NaN         0       crew  
2205        England       NaN    NaN         0        NaN  
2206        England       NaN    NaN         0        NaN  

[2207 rows x 10 columns]

In [94]:
df_passenger = df_titanic[df_titanic["type"] == "passenger"]

df_passenger

name  gender   age class embarked  \
0                Abbing, Mr. Anthony    male  42.0   3rd        S   
1          Abbott, Mr. Eugene Joseph    male  13.0   3rd        S   
2        Abbott, Mr. Rossmore Edward    male  16.0   3rd        S   
3     Abbott, Mrs. Rhoda Mary 'Rosa'  female  39.0   3rd        S   
4        Abelseth, Miss. Karen Marie  female  16.0   3rd        S   
...                              ...     ...   ...   ...      ...   
1314          Yvois, Miss. Henriette  female  24.0   2nd        S   
1315       Zakarian, Mr. Mapriededer    male  22.0   3rd        C   
1316             Zakarian, Mr. Ortin    male  27.0   3rd        C   
1317               Zenni, Mr. Philip    male  25.0   3rd        C   
1318             Zimmermann, Mr. Leo    male  29.0   3rd        S   

            country  ticketno     fare  survived       type  
0     United States    5547.0   7.1100         0  passenger  
1     United States    2673.0  20.0500         0  passenger  
2     United States    2673.0  20.0500         0  passenger  
3           England    2673.0  20.0500         1  passenger  
4            Norway  348125.0   7.1300         1  passenger  
...             ...       ...      ...       ...        ...  
1314         France  248747.0  13.0000         0  passenger  
1315         Turkey    2656.0   7.0406         0  passenger  
1316         Turkey    2670.0   7.0406         0  passenger  
1317        Lebanon    2620.0   7.0406         1  passenger  
1318        Germany  315082.0   7.1706         0  passenger  

[1317 rows x 10 columns]

In [95]:
titanic_ct = pd.crosstab(df_passenger["embarked"], df_passenger["class"])

titanic_ct

class     1st  2nd  3rd
embarked               
B           3    6    0
C         143   26  102
Q           3    7  113
S         175  245  494

## 2. Using the joint distribution that calculated above, calculate the following:

* the conditional distribution of their class given where they embarked
* the conditional distribution of where they embarked given their class

Use the conditional distributions that you calculate to answer the following quesitons:

* What proportion of 3rd class passengers embarked at Southampton?
* What proportion of Southampton passengers were in 3rd class?

In [96]:
df_passenger.groupby("class")["embarked"].value_counts(normalize=True)

class  embarked
1st    S           0.540123
       C           0.441358
       B           0.009259
       Q           0.009259
2nd    S           0.862676
       C           0.091549
       Q           0.024648
       B           0.021127
3rd    S           0.696756
       Q           0.159379
       C           0.143865
Name: proportion, dtype: float64

In [97]:
df_passenger.groupby("embarked")["class"].value_counts(normalize=True)

embarked  class
B         2nd      0.666667
          1st      0.333333
C         1st      0.527675
          3rd      0.376384
          2nd      0.095941
Q         3rd      0.918699
          2nd      0.056911
          1st      0.024390
S         3rd      0.540481
          2nd      0.268053
          1st      0.191466
Name: proportion, dtype: float64

**69.78% of 3rd class passengers are Southhampton
while 54.05% of Southhampton passengers are 3rd class**

This helps create a conditional distribution in a percentage format

In [100]:
titanic_ct.sum(axis=1)
titanic_ct.divide(titanic_ct.sum(axis=1), axis=0)

class          1st       2nd       3rd
embarked                              
B         0.333333  0.666667  0.000000
C         0.527675  0.095941  0.376384
Q         0.024390  0.056911  0.918699
S         0.191466  0.268053  0.540481

## 3. Make a visualization showing the distribution of a passenger's class, given where they embarked.

Discuss the pros and cons of using this visualization versus the distributions you calculated before, to answer the previous questions.

In [ ]:
import plotnine import *

(ggplot())